lib/ddq-backend-mysql.js

```diff
@@ -0,0 +1,374 @@
+"use strict";
+
+const EMIT_DATA = "data", EMIT_ERR = "error";
+
+module.exports = (configValidation, crypto, EventEmitter, mysql, timers) => {
+    var cycleCounter, writeRecord;
+
+    /**
+     * Updates heartbeatDate of a specific record and calls the callback that is
+     * passed in.
+     *
+     * @param {Object} ddqBackendInstance
+     * @param {string} recordId
+     * @param {Function} callback
+     */
+    function heartbeat(ddqBackendInstance, recordId, callback) {
+        ddqBackendInstance.connection.query(`UPDATE ??
+            SET heartbeatDate = NOW()
+            WHERE hash = ?
+                AND isProcessing = true
+                AND owner = ?;`,
+            [
+                ddqBackendInstance.config.table,
+                recordId,
+                ddqBackendInstance.owner
+            ], callback
+        );
+    }
+
+
+    /**
+     * Updates a specific record so that it can be requeued.
+     *
+     * @param {Object} ddqBackendInstance
+     * @param {string} recordId
+     * @param {Function} callback
+     */
+    function requeue(ddqBackendInstance, recordId, callback) {
+        ddqBackendInstance.connection.query(`UPDATE ??
+            SET owner = null,
+                isProcessing = false,
+                requeued = false
+            WHERE hash = ?;`,
+            [
+                ddqBackendInstance.config.table,
+                recordId
+            ], callback
+        );
+    }
+
+
+    /**
+     * Deletes a specific record from the database. If unsuccessful, this will
+     * call the command to requeue the record.
+     *
+     * @param {Object} ddqBackendInstance
+     * @param {string} recordId
+     * @param {Function} callback
+     */
+    function remove(ddqBackendInstance, recordId, callback) {
+        ddqBackendInstance.connection.query(`DELETE FROM ??
+            WHERE hash = ?
+                AND requeued = false
+                AND owner = ?;`,
+            [
+                ddqBackendInstance.config.table,
+                recordId,
+                ddqBackendInstance.owner
+            ], (err) => {
+                if (err) {
+                    requeue(ddqBackendInstance, recordId, callback);
+                }
+
+                callback(err);
```

Comment:
If error, requeue.  If requeue errors, then tell the callback there's a problem.

With this structure, the callback will get called twice when there is any error.  We really don't want that.  So, wrap the lower portion in an `else` to make this all better.